### DISCUSSION TITLE
Is there a way to determine if a package is provided by OS repos or custom built?

### DISCUSSION THREAD

**Discussion by sshedi** [rpm-software-management/rpm#2808] [General]
Hi, I'm not sure if something like this exists already; asking here for inputs.

I'm using an rpm based distro and I want to know what packages are installed from OS's repo and what packages are custom built and installed from different source.

Conditions:
- Closed eco system, no network connection.
- The custom built rpm has same info in the header. Meaning, `Distribution, Vendor etc.,` is in alignment with other rpms.

I can check for signature key id (in `rpm -qi`) but it will add some other type of complexities.

Would be nice to have a provision to add a flag in rpm header, something like:
```
...
Distribution: RHEL
Vendor: Redhat, Inc.
ProvidedByOS: 1 / 0
OR
Provider: OS / 3rd party
...
```
Thoughts?

**Comment by voxik** [comment]
I'd say this is out of scope for RPM, but e.g. DNF can have the answer:

~~~
# dnf info rpm | grep repository
From repository : rawhide
~~~

**Reply by sshedi** [reply]
Thanks for the repsonse. This doesn't work in Fedora-39.
```
$
➜ $ dnf info rpm --disablerepo=* | grep repository
$
```

> I'd say this is out of scope for RPM

Why do you believe so? It should be possible to have some kind of metadata to indicate that the rpm is provided by OS. Not a mandatory field but on the need basis.

**Comment by voxik** [comment]
> Thanks for the repsonse. This doesn't work in Fedora-39.
> 
> ```
> $
> ➜ $ dnf info rpm --disablerepo=* | grep repository
> $
> ```

I was testing on Rawhide with DNF5. With DNF4, the output might be a bit different, but I believe that `dnf info rpm` still works.

> 
> > I'd say this is out of scope for RPM
> 
> Why do you believe so? It should be possible to have some kind of metadata to indicate that the rpm is provided by OS. Not a mandatory field but on the need basis.

Because RPM does not have concept of something like OS in your view. RPM does not even have a concept of e.g. repository, which IMHO is narrower concept then OS

**Comment by pmatilai** [comment]
The package signature is the strongest indicator of origin there is. Otherwise it gets very very gray: if you download a package from a repo and then install locally, the connection to the repo was lost. Or, if you copy a package from one repo to another, the package does not change but the repo does.
